govcd/edgegateway.go

```diff
@@ -200,82 +200,75 @@ func (eGW *EdgeGateway) RemoveNATPortRule(networkHref, natType, externalIP, exte
-	splitPath := strings.Split(cleanPath, "/")
-
-	if len(splitPath) < 2 {
-		return "", fmt.Errorf("path does not contain url path: %s", splitPath)
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
 	}
 
-	objectID := splitPath[len(splitPath)-1]
-
-	return objectID, nil
-}
-
-func (eGW *EdgeGateway) RemoveNATMapping(natType, externalIP, internalIP, port string) (Task, error) {
-	return eGW.RemoveNATPortMapping(natType, externalIP, port, internalIP, port)
+	return nil
 }
 
-func (eGW *EdgeGateway) RemoveNATPortMapping(natType, externalIP, externalPort, internalIP, internalPort string) (Task, error) {
-	// Find uplink interface
-	var uplink types.Reference
-	for _, gi := range eGW.EdgeGateway.Configuration.GatewayInterfaces.GatewayInterface {
-		if gi.InterfaceType != "uplink" {
-			continue
-		}
-		uplink = *gi.Network
+// RemoveNATRuleAsync removes NAT rule and return  or error.
+// Old functions RemoveNATPortMapping and RemoveNATMapping
+// removes using rule details and expects interface is external network.
+func (eGW *EdgeGateway) RemoveNATRuleAsync(id string) (Task, error) {
+	if "" == id {
+		return Task{}, fmt.Errorf("provided id is empty")
 	}
 
-	newEdgeConfig := eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration
-
-	// Take care of the NAT service
-	newNatService := &types.NatService{}
-
-	newNatService.IsEnabled = newEdgeConfig.NatService.IsEnabled
-	newNatService.NatType = newEdgeConfig.NatService.NatType
-	newNatService.Policy = newEdgeConfig.NatService.Policy
-	newNatService.ExternalIP = newEdgeConfig.NatService.ExternalIP
-
-	for _, natRule := range newEdgeConfig.NatService.NatRule {
+	err := eGW.Refresh()
+	if err != nil {
+		return Task{}, fmt.Errorf("error refreshing edge gateway: %#v", err)
+	}
 
-		if natRule.RuleType == natType &&
-			natRule.GatewayNatRule.OriginalIP == externalIP &&
-			natRule.GatewayNatRule.OriginalPort == externalPort &&
-			natRule.GatewayNatRule.Interface.HREF == uplink.HREF {
-			util.Logger.Printf("[DEBUG] REMOVING %s Rule: %#v", natRule.RuleType, natRule.GatewayNatRule)
-			continue
+	natServiceToUpdate := eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService
+	ruleIndex := -1
+	if natServiceToUpdate != nil {
+		for n, existingNatRule := range natServiceToUpdate.NatRule {
+			if existingNatRule.ID == id {
+				ruleIndex = n
+				break
+			}
 		}
-		util.Logger.Printf("[DEBUG] KEEPING %s Rule: %#v", natRule.RuleType, natRule.GatewayNatRule)
-		newNatService.NatRule = append(newNatService.NatRule, natRule)
+	} else {
+		return Task{}, fmt.Errorf("edge gtw doesn't have NAT rules")
 	}
 
-	newEdgeConfig.NatService = newNatService
+	if ruleIndex == -1 {
+		return Task{}, fmt.Errorf("edge gtw doesn't have rule with such Id")
+	}
+
+	natServiceToUpdate.NatRule = append(natServiceToUpdate.NatRule[:ruleIndex], natServiceToUpdate.NatRule[ruleIndex+1:]...)
```

Comment:
Wouldn't this fail on incorrect index if we had only 1 rule (array element 0)?

Reply:
didn't fail. Added check to be 100% valid

Reply:
But in this case having exactly one rule would miss it. Wouldn't it?

Reply:
From running point works fine. Also added check for 1 rule case

Reply:
Ok. Got it.